Introduce tiers.md (#242)
meilisearch/integration-guides: resources/tiers.md

# SDK tiers

The objective of Meilisearch's Integrations Team is to **ease Meilisearch usage to as many developers as possible**.

> Given this statement, **tiers** are nothing more than a __group of SDKs/integrations__ where the integrations team can fairly divide their attention. Unfortunately is not possible to keep all the SDKs 100% updated against the new Meilisearch releases all the time, the Integrations Team is a polyglot small team that maintains 30+ different integrations and tools, so it is humanly impossible to consistently meet the Meilisearch quality standards.

Reviewer: ```suggestion
Given this statement, **tiers** are nothing more than a __group of SDKs/integrations__ where the integrations team can fairly divide their attention. Unfortunately is not possible to keep all the SDKs 100% updated against the new Meilisearch releases all the time. The Integrations Team is a polyglot small team that maintains 30+ different integrations and tools, so it is humanly impossible to consistently meet the Meilisearch quality standards.
```

> The ultimate goal of introducing tiers is to allow the team to have more time for new projects/integrations so that more time can be invested and more users will be impacted.

Reviewer: ```suggestion
The ultimate goal of introducing tiers is to allow the team to have more time to invest in new projects/integrations in order to impact more users.
```

> It's been using [data collected here](https://docs.meilisearch.com/learn/what_is_meilisearch/telemetry.html) and data like GitHub stars, forks, and watchers to measure the impact of an SDK.

Reviewer: What is `it` ?

Author: Yeah, it was bad translation here (I wrote the article initially in pt-br 😝). I changed the whole phrase to:

To measure the impact of an SDK, the data came from different places like [telemetry](https://docs.meilisearch.com/learn/what_is_meilisearch/telemetry.html) and GitHub's stars, forks, and watchers.

> It is worth remembering that this priority is just a concept introduced by the team where it aims to give a decision weight before
> we invest time in some issues of some SDK in question.

Reviewer: ```suggestion
It is worth remembering that this priority list is just a concept introduced by the team aiming to weigh a task before investing time on it. 
```

Not sure I understand the sentence correctly.

Author: It means nothing is written in stone. It does not mean that we never solve an issue in a `tier #3` SDK if we have an open issue in `tier #2` for example.

The idea here is to give the reader the sense that we're going to use this concept for prioritization purposes and not to block the road for impactless SDKs. 

It is clearer now? Maybe I should change the way I wrote it... Do you have any suggestion now?

Author: WDYT?

> It is worth remembering that this priority is just a concept introduced by the team that aims to give a decision weight before investing time in some issues of the SDKs. So it does not mean we will never work on feature requests from the bottom tiers anymore, but it says we will prioritize the top tiers most of the time transparently.

Reviewer: Good for me!

> :warning: We will keep all officially maintained SDKs **working**. :warning:

Reviewer: ```suggestion
:warning: We keep all officially maintained SDKs **working**. :warning:
```

What do you mean with working? We do not leave bugs behind?

Author: Not really. Let me give you a hypothetical example of it:

Let's take my scope, for example, if a user reports an error in the Dart like this https://github.com/meilisearch/meilisearch-dart/issues/242.  
I'll not leave them alone to work on a `feature request` of PHP.

Is it clear?

Reviewer: Okey! Yes it is :)  We fix critical bugs

## Tier #1

> The SDKs in group #1 have total priority over the team's time. Issues are visited before issues are present in tier #2, for example.

Reviewer: ```suggestion
The SDKs in group #1 have total priority over the team's time. For example, issues are visited before issues are present in tier #2.
```

It is worth remembering that this does not mean open issues will be finalized promptly. Every issue needs time to be understood/confirmed/worked on, and consequently closed.

> All new Meilisearch features that make sense for those SDKs will be available in the release day.

Reviewer: ```suggestion
All new Meilisearch features that make sense for those SDKs will be available on the release day.
```

## Tier #2

> Tier #2 SDKs are the emergent ones, we strongly believe they can move further, and we are betting on them.

Reviewer: Like parana in serie B.

Author: 🤣🤣🤣🤣🤣🤣🤣🤣

> In any case, since they do not bring that much value compared to tier #1 we will only apply the amount of time required to keep it working and do important maintenance code and code cleanup.

Reviewer: ```suggestion
In any case, since they bring less value compared to tier #1 we only spend the amount of time required to keep it working and avoid important maintenance code and code cleanup.
```

What do you mean with `important maintenance` ?

Author: Actually, I'm referring to things the SDK should have, but they are not impacting all the users. It is harder to give a clear example, but I want to mean that we still can use some time on non-critical issues, you know?

> Only in urgent cases where, for example, the SDK is no longer working issues from the SDKs in group #2 will be prioritized in comparison to tier #1.

Reviewer: ```suggestion
Only in urgent cases where, for example, the SDK is no longer working, issues from the SDKs in group #2 are prioritized over tier #1.
```

## Tier #3

The SDKs in group #3 have relatively low user adoption compared to groups #2 and #1. They are probably new or have a small community behind the tech stack in the first place.

> As with tier #2 SDKs in this group will be prioritized only in critical cases, and they are not going to receive active improvements from the integrations team.

Reviewer: ```suggestion
As with tier #2, SDKs in this group are prioritized only in critical cases, and they are not going to receive active improvements or maintance from the integrations team.
```

Active improvements?

Author: You know, after the introduction of the tiers, I'll not spend time going through the Swift SDK looking for ways of improving the code, since I can do it in Ruby (which is tier `#2`), or even in the PHP `#1`.

This is a very special group because we really need help from the community here.

## Tiers list

Integration | Tier |
-------------|------|
Javascript | #1 |
PHP | #1 |
Instant Meilisearch | #1 |
Python | #1 |
Go | #1 |
Strapi | #2 |
Ruby | #2 |
Rails | #2 |
.NET | #2 |
Rust | #2 |
Symfony | #3 |
docs-searchbar.js | #3 |
Firebase | #3 |
Dart | #3 |

> Java | #3 |

Reviewer 2: Why Java is only on tier #3?

Reviewer 3: I think you indeed should keep an eye on Java usage, because the community is huge and already came to a lot of time (they are currently a little bit frustrated because our SDK is not up to date). Can quickly become tier #2 or (even tier #1 😱)

Author: I hope they will! But for now, since we don't have data coming from the Java users, Java is `#3`. I'll share what is my rationale behind the list and the order after merging this first version :)

Swift | #3 |
Vuepress | #3 |
Gatsby | #3 |

## FAQ

> ### How will the tiers be organized?

Reviewer: ```suggestion
### How are the tiers organized?
```

The method used is mostly manual and subjective, where we use the data previously collected from the [telemetry](https://docs.meilisearch.com/learn/what_is_meilisearch/telemetry.html) and Github to generate groups.

> At first there are only 3 tiers, and in the future this number may change.

Reviewer: ```suggestion
To begin we only have 3 tiers, in the future, this number may change.
```

> ### Is it possible for an SDK to be moved up a level?

Reviewer: ```suggestion
### Is it possible for an SDK to move up a level?
```

> Yes, fully possible! Especially if the team detects that more users are interested and definitely using it.

Reviewer: ```suggestion
Yes, fully possible! Especially if the team detects that more users are interested in and using it.
```

> Whenever a anomaly is detected, a change can be submitted to the `integration-guides` and the groups will be modified to better manage the team's time. The reassessment of the tiers will be carried out every release cycle (8 weeks).

Reviewer: > The reassessment of the tiers will be carried out every release cycle (8 weeks).

What do you mean with that?

Author: The tiers will change after some time (at least, I expect that). For example, we could provide a UI Kit for Flutter developers who will make Dart a lot more important than it is, and this can increase its importance + usage, so it makes sense to move it to the top tiers.

> ### How to identify that an SDK belongs to some level?

Reviewer: ```suggestion
### How to know the tier list on an SDK?
```

Author: WDYT? How to identify which tier an SDK belongs to?

> It is possible by accessing the README.md of each repository, and it will contain a tier badge
> below the Meilisearch logo, this badge says: "tiers #1", "tiers #2" and "tiers #3" like this:

Reviewer: ```suggestion
It is possible by accessing the README.md of each repository. A tier badge is present below the Meilisearch logo. The badge can have the following values: "tiers #1", "tiers #2" and "tiers #3". For example:
```

[![tiers - #1](https://img.shields.io/static/v1?label=tiers&message=%231&color=blue)](https://github.com/meilisearch/integration-guides/blob/main/resources/tiers.md)
[![tiers - #2](https://img.shields.io/static/v1?label=tiers&message=%231&color=2ea44f)](https://github.com/meilisearch/integration-guides/blob/main/resources/tiers.md)
[![tiers - #3](https://img.shields.io/static/v1?label=tiers&message=%231&color=critical)](https://github.com/meilisearch/integration-guides/blob/main/resources/tiers.md)

#### Why SDKs are present in a particular tier?

> Because those SDKs have the most active users whether in the Meilisearch Cloud or not. Also they have a larger or broader community in general. Other principles could be used to move SDKs around if we think is the right thing to do.

Reviewer: ```suggestion
Because those SDKs have the most active users. Not only from users using the Meilisearch Cloud. They have a larger or broader community in general. Other principles could be used to move SDKs around if we think it is relevant.
```

Not sure about `Not only from users using the Meilisearch Cloud.`

> larger or broader

Is there a difference between both these words?

Author: When I share the rationale behind the selection of the tier, it will make more sense. But giving you a spoiler the SDKs who have more cloud customers are rewarded with more "points" than SDKs with fewer customers.

## Tier #2

Tier #2 SDKs are the emergent ones. The Integrations Team believes they can move further and are betting on them.
In any case, since they bring less value than tier #1, the team will only spend the time required to keep these SDKs working and do the necessary maintenance code and code cleanup.

Only in urgent cases where, for example, the SDK is no longer working issues from the SDKs in group #2 prioritized over Tier #1.

## Tier #3

The SDKs in group #3 have relatively low user adoption compared to groups #2 and #1. They are probably new or have a small community behind the tech stack in the first place.
As with tier #2, SDKs in this group are prioritized only in critical cases, and they are not going to receive active improvements or maintance from the integrations team.

This is a very special group because the Integrations Team really need help from the community here.

## Tiers list

Integration | Tier |
-------------|------|
Javascript | #1 |
PHP | #1 |
Instant Meilisearch | #1 |
Python | #1 |
Go | #1 |
Strapi | #2 |
Ruby | #2 |
Rails | #2 |
.NET | #2 |
Rust | #2 |
Symfony | #3 |

> docs-searchbar.js | #3 |

Reviewer 4: Is there a reason to have `docs-searchbar.js` being tier 3? 

That's potentially something we will want to maintain / replace actively at some point because It's an excellent opener to have a tool that works to get many projects moving from Algolia to Meilisearch for their documentation.
 
What's the rationale for prioritizing a severe issue on tier 3?

Author: Not really. It is just because we don't have the User Agent there!

Firebase | #3 |
Dart | #3 |
Java | #3 |
Swift | #3 |
Vuepress | #3 |
Gatsby | #3 |

## FAQ
### How are the tiers organized?

The method used is mainly manual and subjective, using previously collected data from the [telemetry](https://docs.meilisearch.com/learn/what_is_meilisearch/telemetry.html) and Github to generate the groups.

Only 3 tiers were defined, but this number may change in the future, and this guide will be updated accordindgly.

### Is it possible for an SDK to move up a level?

Yes, fully possible! Especially if the team detects that more users are interested in and using it.
Whenever a anomaly is detected, a change can be submitted to the `integration-guides` and the groups will be modified to better manage the team's time. The reassessment of the tiers will be carried out every release cycle (8 weeks).

The other way around is also valid. A SDK can move to less important tiers. :shipit:

### How to identify which tier an SDK belongs to?

> It is possible by accessing the README.md of each repository. A tier badge is present below the Meilisearch logo. The badge can have the following values: "tiers #1", "tiers #2" and "tiers #3". For example:

Reviewer: I'm not convinced that adding the badges is a good idea. SDK's will still work with the latest versions of Meilisearch and most users will be satisfied with its current state. 

By warning them that this is a low priority SDK it might result in:
- User  fearing deprecation
- User complaining because they don't want their SDK in that tier list even if they would never user the missing features
- Give the impression that we are going backward in activity like a "cut in budget"

If something is missing, an issue exist in any case with the missing feature which reassure the user that we did not forget about the SDK, provide them a place to tell us they need it, and help them implement it if they want it sooner than later.

[![tiers - #1](https://img.shields.io/static/v1?label=tiers&message=%231&color=blue)](https://github.com/meilisearch/integration-guides/blob/main/resources/tiers.md)

> [![tiers - #2](https://img.shields.io/static/v1?label=tiers&message=%231&color=2ea44f)](https://github.com/meilisearch/integration-guides/blob/main/resources/tiers.md)

Reviewer 4: ```suggestion
[![tiers - #2](https://img.shields.io/static/v1?label=tiers&message=%232&color=2ea44f)](https://github.com/meilisearch/integration-guides/blob/main/resources/tiers.md)
```

Author: Good catch! CTRL+C/CTRL+V spotted! 🙈

> [![tiers - #3](https://img.shields.io/static/v1?label=tiers&message=%231&color=critical)](https://github.com/meilisearch/integration-guides/blob/main/resources/tiers.md)

Reviewer 4: ```suggestion
[![tiers - #3](https://img.shields.io/static/v1?label=tiers&message=%233&color=critical)](https://github.com/meilisearch/integration-guides/blob/main/resources/tiers.md)
```

Author: Good catch! CTRL+C/CTRL+V spotted! 🙈
